ccinput/tests/test_nwchem.py

```diff
@@ -0,0 +1,1430 @@
+            "file": "Cl.xyz",
+            "software": "nwchem",
+            "charge": "-1",
+            "method": "B3LYP",
+            "basis_set": "6-31+G(d,p)",
+        }
+
+        inp = self.generate_calculation(**params)
+
+        REF = """
+        TITLE "File created by ccinput"
+        start Cl
+        memory total 10000 mb
+        charge -1
+
+        geometry units angstroms
+        Cl   0.00000000   0.00000000   0.00000000
+        end
+
+        basis
+        * library 6-31+G*
+        end
+
+        dft
+        xc b3lyp
+        mult 1
+        end
+
+        task dft freq
+
+
+        """
+
+        self.assertTrue(self.is_equivalent(REF, inp.input_file))
+
+    def test_freeze_bond_DFT(self):
+        params = {
+            "nproc": 8,
+            "mem": "10000MB",
+            "type": "Constrained Optimisation",
+            "file": "ethanol.xyz",
+            "software": "nwchem",
+            "charge": "0",
+            "method": "B3LYP",
+            "basis_set": "6-31+G(d,p)",
+            "constraints": "Freeze/1_2;",
+        }
+
+        inp = self.generate_calculation(**params)
+
+        REF = """
+        TITLE "File created by ccinput"
+        start ethanol
+        memory total 10000 mb
+        charge 0
+
+        geometry units angstroms
+        C   -1.31970000  -0.64380000   0.00000000
+        H   -0.96310000  -1.65260000   0.00000000
+        H   -0.96310000  -0.13940000  -0.87370000
+        H   -2.38970000  -0.64380000   0.00000000
+        C   -0.80640000   0.08220000   1.25740000
+        H   -1.16150000   1.09160000   1.25640000
+        H   -1.16470000  -0.42110000   2.13110000
+        O    0.62360000   0.07990000   1.25870000
+        H    0.94410000   0.53240000   2.04240000
+        end
+
+        basis
+        * library 6-31+G*
+        end
+
+        dft
+        xc b3lyp
+        mult 1
+        end
+
+        constraints
+        spring bond 1 2
+        end
+
+        task dft optimize
+
+
+        """
+
+        self.assertTrue(self.is_equivalent(REF, inp.input_file))
+
+    def test_freeze_angle_DFT(self):
+        params = {
+            "nproc": 8,
+            "mem": "10000MB",
+            "type": "Constrained Optimisation",
+            "file": "ethanol.xyz",
+            "software": "nwchem",
+            "charge": "0",
+            "method": "B3LYP",
+            "basis_set": "6-31+G(d,p)",
+            "constraints": "Freeze/2_1_3;",
+        }
+        with self.assertRaises(UnimplementedError):
+            self.generate_calculation(**params)
+
+    def test_invalid_opt_mod(self):
+        params = {
+            "nproc": 8,
+            "mem": "10000MB",
+            "type": "Constrained Optimisation",
+            "file": "ethanol.xyz",
+            "software": "nwchem",
+            "charge": "0",
+            "method": "B3LYP",
+            "basis_set": "6-31+G(d,p)",
+            "constraints": "",
+        }
+
+        with self.assertRaises(Exception):
```

Comment:
Instead of `Exception` here, `InvalidParameter` would be better. It is more specific and would make the test fail if there is some other exception in the code. This also applies to the other expected failures below.